---
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
    theme: flatly
params:
  set_title: "My Title!"
  pair: "pair.id"
  jabba_rds: "./jabba.simple.rds"
  outdir: "./"
  tumor_type: "UNKNOWN"
  server: "https://mskilab.com/gGraph/"
  del_thresh: 0.5
  amp_thresh: 4
---

<style>
  .superbigimage{
  overflow-x:scroll;
  white-space: nowrap;
  }
  
  .superbigimage img{
  max-width: none;
  }
</style>

---
title: Whole genome/transcriptome sequencing of `r params$pair`, `r
paste(ifelse(grepl("^[aeiou]", params$tumor_type, ignore.case = T),
"an", "a"), params$tumor_type)`
---

```{r, setup, include = FALSE, echo = FALSE, message = TRUE}
## global code chunk option
knitr::opts_chunk$set(
    collapse = TRUE,
    fig.width = 8,
    fig.height = 8,
    message = FALSE,
    warning = FALSE,
    echo = FALSE)

## hopefully make skidb standard asap
library(gGnome)
library(gTrack)
library(skidb)
library(ggplot2)
library(ggforce)
library(kableExtra)
library(hwriter)
library(htmlwidgets)
library(svglite)
library(slickR)
library(plotly)
library(shiny)
message("Loaded Packages")

gg = readRDS(paste0(params$outdir, "/complex.rds"))
jb = readRDS(params$jabba_rds)
cvgt = readRDS(paste0(params$outdir, "/coverage.gtrack.rds"))
jj = gg$junctions[type=="ALT"]
message("Loaded gGraph")
```

***
# `r paste(params$pair, params$tumor_type, sep = ", ")`

This is a `r params$tumor_type` sample. Average read depth for the tumor
sample is xx. Estimated purity is 
`r format(jb$purity, digits = 2)` 
and ploidy `r format(jb$ploidy, digits = 2)`. This sample has likely
undergone whole genome doubling (WGD) xx time(s). xx percent of the
genome shows copy number alterations (CNA, defined as CN < `r params$del_thresh` or
>= `r params$amp_thresh` times the ploidy). 

Over the whole genome, the number of somatic SNV/INDEL is xx, or Tumor Mutation Burden (TMB) of xx per Mbp. The total junction burden is `r length(jj)`.

***
# Whole genome view {.tabset .tabset-fade .tabset-pills}
## gTrack 
<div class="superbigimage">
```{r, wgs-gtrack, max.width = "95%", results = "asis", echo = FALSE, fig.align = "center"}
knitr::include_graphics(file.path(params$outdir, "wgs.gtrack.png"))
```
</div>

## circos

```{r, wgs-circos, out.width = "95%", results = "asis", echo = FALSE, fig.align = "center"}
knitr::include_graphics(file.path(params$outdir, "wgs.circos.png"))
```

<br>


***

# Copy number variants
Copy number variations (CNVs) in oncogenes and tumor suppressor genes.

```{r, driver-cnv-table, results = "asis", max.width= "100%"}
driver.genes.cnv = fread(file.path(params$outdir, "driver.genes.cnv.txt"))
## driver.genes.cnv[cnv != ''] %>% kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "300px")
DT::datatable(driver.genes.cnv, options = list(scrollX = TRUE))
```


## Gallery {.tabset .tabset-fade .tabset-pills}

```{r cn-tabset, results = "asis", echo = FALSE, out.width= "100%", fig.align = "center"}
cn.gallery.dt = fread(file.path(params$outdir, "cn.gallery.txt"))
if (nrow(cn.gallery.dt) > 0) {
    cat("\n")
    for (g in cn.gallery.dt[, gene_name]) {
        cat("\n")
        cat("###", paste0(g, "<br>","\n"))
	cat(
		paste0('<p>',
			'<a href="',
			cn.gallery.dt[gene_name == g, plot.link],
			'">',
			'<img src="',
			cn.gallery.dt[gene_name == g, plot.fname],
			'" width="100%" />',
			'</a>',
			'</p>'
		)
	)
        cat('\n', '<br>', '\n\n')
    }
}
```

***

# Fusions

## Known fusions
Fusions in which at least one involved gene is an oncogene or tumor suppressor.

```{r, driver-fusions-table, results = "asis", out.width= "100%"}
driver.fusions.fname = file.path(params$outdir, "fusions.driver.txt")
if (file.exists(driver.fusions.fname)) {
   driver.fusions.dt = fread(driver.fusions.fname)
   driver.fusions.cols = intersect(c("genes", "walk.id", "driver.name", "chroms", "maxcn", "total.aa", "gene.pc", "ev.id", "ev.type"), colnames(driver.fusions.dt))
   dat = driver.fusions.dt[, ..driver.fusions.cols]
   DT::datatable(dat, options = list(scrollX = TRUE))
}
```

## Gallery {.tabset .tabset-fade .tabset-pills}

```{r fusion-tabset, results = "asis", echo = FALSE, out.width= "100%", fig.align = "center"}
fusions.gallery.dt = fread(file.path(params$outdir, "fusions.driver.txt"))
if (nrow(fusions.gallery.dt) > 0) {
    cat("\n")
    for (g in fusions.gallery.dt[, genes]) {
        cat("\n")
        cat("###", paste0(g, "<br>","\n"))
	cat(
		paste0('<p>',
			'<img src="',
			fusions.gallery.dt[genes == g, plot.fname],
			'" width="75%" />',
			'</p>'
		)
	)
        cat('\n', '<br>', '\n\n')
    }
}
```

## Other in-frame, non-silent gene fusions

```{r, other-fusions-table, results = "asis", max.width= "100%"}
other.fusions.fname = file.path(params$outdir, "fusions.other.txt")
if (file.exists(other.fusions.fname)) {
   other.fusions.dt = fread(other.fusions.fname)
   other.fusions.cols = intersect(c("genes", "walk.id", "driver.name", "chroms", "maxcn", "total.aa", "gene.pc", "ev.id", "ev.type"), colnames(driver.fusions.dt))
   dat = other.fusions.dt[, ..other.fusions.cols]
   DT::datatable(dat, options = list(scrollX = TRUE))
}
```

## Gallery

```{r, other-fusions-gallery, results = "asis", echo = FALSE, out.width= "100%", fig.align = "center"}

if (file.exists(other.fusions.fname)) {
   other.fusions.dt = fread(other.fusions.fname)
   if (nrow(other.fusions.dt) > 0) {
      slick_up = slickR(obj = other.fusions.dt$plot.fname, height = 500, width = "95%") + settings(slidesToShow = 1, slidesToScroll = 1) + settings(dots = TRUE)
slick_up
	}
}
```


## Altered transcripts
Coming soon

Under construction

***

# Over/under-expression of oncogenes/TSGs 
```{r, driver-expr-table, results = "asis", max.width= "100%"}
driver.genes.expr = fread(file.path(params$outdir, "driver.genes.expr.txt"))
DT::datatable(driver.genes.expr, options = list(scrollX = TRUE))
```

## Waterfall plot
```{r, driver-waterfall, results = "asis", max.width = "100%"}
waterfall.fn = file.path(params$outdir, "waterfall.png")
if (file.exists(waterfall.fn)) {
   knitr::include_graphics(waterfall.fn)
}
```

## Gallery {.tabset .tabset-fade .tabset-pills}

```{r expression, results = "asis", echo = FALSE, out.width= "100%", fig.align = "center"}
if (file.exists(paste0(params$outdir, "/cool.expr.rds"))){
    cool.exp = readRDS(paste0(params$outdir, "/cool.expr.rds"))
    expr.slickr.dt = fread(file.path(params$outdir, "expr.gallery.txt"))
    cat("\n")
    for (g in cool.exp[order(qt, decreasing = T), gene]){
        cat("\n")
        cat("###", cool.exp[g, paste0(gene, " (", direction, ")", "<br>", "\n")])
        ## this.png = try(readPNG(file.path(
        ##     params$outdir, cool.exp[g, paste0("/", gene, ".", direction, ".expr.png")]
        ## )))
        ## try(grid::grid.raster(this.png,width = 1.0, just = "center"))
	## try just pasting the html to get side by side
	cat(
		paste0(
			'<p float="left">',
			'<img src="',
			params$outdir,
			cool.exp[g, paste0("/", gene, ".", direction, ".expr.png")],
			'" width="45%" />',
			'<img src="',
			expr.slickr.dt[gene_name == g, plot.fname],
			'" width="45%" />',
			'</p>'
		)
	)
        ## cat(
        ##     paste0(
        ##         "![](",
        ##         params$outdir,
        ##         cool.exp[g, paste0("/", gene, ".", direction, ".expr.png")],
        ##         ")",
	## 	"![](",
	## 	expr.slickr.dt[gene_name == g, plot.fname],
	## 	")"),
        ##     "\n")
        ## knitr::include_graphics()
        cat('\n', '<br>', '\n\n')
    }
}
```

***

# SV event classifications
## Burdens of event types
Count of complex events against background distribution from Cell cohort.

```{r, sv-burden, fig.align = "center", out.width = "80%"}
knitr::include_graphics(file.path(params$outdir, "ridgeplot.png"))
```

## Gallery of events

```{r, SV, results = "asis", echo = FALSE, out.width= "100%", fig.align = "center"}
## read data table with plot png/URL
sv.slickr.dt = fread(file.path(params$outdir, "sv.gallery.txt"))[!is.na(plot.link),]
slick_up = slickR(obj = sv.slickr.dt$plot.fname, height = 500, width = "95%", objLinks = sv.slickr.dt$plot.link) + settings(slidesToShow = 1, slidesToScroll = 1) + settings(dots = TRUE)
slick_up
## slick_down = slickR(obj = sv.slickr.dt$plot.fname, height = 100, width = "95%", objLinks = sv.slickr.dt$plot.link) + settings(slidesToScroll = 1,  slidesToShow = 3, centerMode = TRUE, focusOnSelect = TRUE)
## slick_up %synch% slick_down
```

# SNV signatures {.tabset .tabset-fade .tabset-pills}
Fitting COSMIC signatures to the whole genome mutation profiles of
this sample we obtain the following active SNV mutation signatures:

```{r, SNV signature, results = "asis", echo = FALSE, out.width="100%", fig.align = "center"}
cat("\n")
cat("## deconstructSigs: input profile, fitted, error")
cat("\n")
## knitr::include_graphics(paste0(params$outdir, "/deconstruct_sigs.png"))

cat(
    paste0("![](", paste0(params$outdir, "/deconstruct_sigs.png"),")"), "\n"
)

cat ("\n<br>\n")

cat("\n")
cat("## SNV signature composition")
cat("\n")

if (file.exists(paste0(params$outdir, "/sig.composition.png"))) {
   cat(
	 paste0("![](", paste0(params$outdir, "/sig.composition.png"),")"), "\n")
}	 

cat ("\n<br>\n")
```

```{r, HRDetect, results = "asis", echo = FALSE, out.width="100%", fig.align = "center"}
if (file.exists(paste0(params$outdir, "/hrdetect.rds"))){
    cat("# HRDetect prediction (homologous recombination deficiency)\n")
    cat("Applying the [HRDetect](https://doi.org/10.1038/nm.4292) model, we
predict homologous recombination deficiency status of this sample
based on its SNV/INDEL/SV signatures.\n\n")

    hrd = readRDS(paste0(params$outdir, "/hrdetect.rds"))
    hrd.yes = hrd[variable=="Probability", value>0.7]
    cat("This sample is classified as", ifelse(hrd.yes, "**HR deficient**", "**NOT** HR deficient"), ".\n")
    
    hrdetect.dims = setNames(        
        c('intercept',
          'Proportion of DELs\nwith microhomology',
          'SBS3',
          'SV signature 3',
          'SV signature 5',
          'HRD index (LOH)',
          'SBS8',
          'Probability of HR Deficiency'),
        c('Intercept', 'del.mh.prop', 'SNV3', 'SV3', 'SV5', 'hrd', 'SNV8', 'Probability')
    )
    
    cat(kbl(hrd[, .(Variable = hrdetect.dims[variable], value = sprintf("%f", value))]) %>%
        kable_classic(full_width = FALSE) %>%
        row_spec(8, bold = T, color = "white", background = ifelse(hrd.yes, "#D7261E", "#1f78b4")))

    cat("\n<br>\n")
    ## insert the plots of input data
    cat(
        paste0("![](", paste0(params$outdir, "/hrdetect.log.dat.png"),")"), "\n\n"
    )
    cat(
        paste0("![](", paste0(params$outdir, "/hrdetect.prop.dat.png"),")"), "\n\n"
    )
}
```

***
# Enhancer hijacking candidates {.tabset .tabset-fade .tabset-pills}
The shortest path on a genome graph provides a lower bound on the
actual genomic distance between two locations in a rearranged
genome. Sometimes rearrangements can bring distal enhancers near a
target gene and activate its expression, thus named _enhancer
hijacking_. We identify such pairs of gene-enhancer where they are >5Mbp
away in the reference genome while potentially <1Mbp in the rearranged genome.

```{r, enhancer hijacking, results = "asis", echo = FALSE, out.width="100%", fig.align = "center"}
prox.plots = dir(params$outdir, "proximity.png$", full.names = TRUE)
if (length(prox.plots)>0){
    for (prox.plot in prox.plots){
        g = gsub(".proximity.png", "", file.name(prox.plot))
        cat("## ", g, "\n<br>\n")
        cat(
            paste0(
                '<p float="left">',
                '<img src="',
                prox.plot,
                '" width="45%" />',
                '<img src="',
                params$outdir,
                cool.exp[g, paste0("/", gene, ".", direction, ".expr.png")],
                '" width="45%" />',
                '</p>'
            )
        )
        ## cat(
        ##     paste0("![](", prox.plot,")"), "\n\n"
        ## )
        cat('\n', '<br>', '\n\n')
    }
}
```
***
# Loose ends
There are xx high quality loose ends in this sample.

***
# Purity and ploidy estimation

***
# Junction read support
Below we show the junction supporting reads.


